torch_geometric/nn/conv/hgt_conv.py

```diff
@@ -139,38 +174,211 @@ def forward(
             In case a node type does not receive any message, its output will
             be set to :obj:`None`.
         """
-
+        xs = list(x_dict.values())
+        if not self.use_gmm:
+            # for segment_matmul check if need padding
+            if self.no_pad:
+                x = torch.cat(xs)
+            else:
+                if self.infer_shapes:
+                    self.dims = torch.tensor(
+                        [x_j.shape[-1] for x_j in x_dict.values()])
+                    self.max_channels = self.dims.max()
+                    input_to_init_w = xs[self.dims.argmax()]
+                    for node_type in x_dict.keys():
+                        self.k_lin[node_type].initialize_parameters(
+                            None, input_to_init_w)
+                        self.q_lin[node_type].initialize_parameters(
+                            None, input_to_init_w)
+                        self.v_lin[node_type].initialize_parameters(
+                            None, input_to_init_w)
+                    self.infer_shapes = False
+                    self.no_pad = (self.dims == self.max_channels).all()
+                x = torch.cat(pad_list(xs, self.dims))
+        elif self.infer_shapes:
+            # initialize lazy params
+            for node_type, x_n in x_dict.items():
+                self.k_lin[node_type].initialize_parameters(None, x_n)
+                self.q_lin[node_type].initialize_parameters(None, x_n)
+                self.v_lin[node_type].initialize_parameters(None, x_n)
+            self.infer_shapes = False
         H, D = self.heads, self.out_channels // self.heads
 
         k_dict, q_dict, v_dict, out_dict = {}, {}, {}, {}
+        # parralelize over node-types
```

Comment:
I feel there is a lot of repetitive logic here. Isn't that something that `HeteroLinear` should be handling (or `HeteroDictLinear`?

Reply:
yes will do

Reply:
done

Reply:
done